### DISCUSSION TITLE
Recommended way to migrate events between event stores

### DISCUSSION THREAD

**Discussion by ahoneggar-trellance** [rasmus/EventFlow#1066] [Q&A]
Our project is on v0 using EventStore for event sourcing and Mongo for read models. We would like to migrate to Mongo for event sourcing as well, and then ultimately migrate to v1. What would be the easiest way to go about the event sourcing migration?

**Comment by rasmus** [comment]
Let's get the easiest out of the way first. Migrating from v0 to v1 in regard to data, is a no-op https://geteventflow.net/migrations/v0-to-v1/ as the data format doesn't change.

Regarding moving from one data store to another, there sadly isn't a tested method. Depending on your system, it could be many GBs or TBs worth of data. I can only give you my thoughts as this isn't something I have considered before, but it certainly is a valid case.

Consider if you need to be live when the migration is being done. If you need to be live, you could potentially solution would be for you to create your own implementation of `IEventPersistence` that wraps both your current EventStore store and your future MongoDB store. Then whenever an aggregate is loaded it is automatically moved from one store to the other (if not already there) and then only committed to MongoDB (or both if you want to be able to gradually move).

Is the wish to move to MongoDB due to the fact that EventStore isn't migrated to v1 yet? If so, then it might be easier and significantly more safe for you if I (or yourself) create a PR to get it moved to v1.

**Reply by ahoneggar-trellance** [reply]
> Consider if you need to be live when the migration is being done.

We do not need to be live, we have regularly scheduled maintenance windows where this could be done "offline".

> Is the wish to move to MongoDB due to the fact that EventStore isn't migrated to v1 yet?

No, the ultimate wish is to part ways with EventStore due to the high resource overhead of having a dedicated server compared to the relatively low event traffic it receives. We've been facing issues with resource spikes when EventStore restarts and re-indexes which requires way more resources compared to standard runtime operation.

Ultimately, I was looking for guidance in replaying events from 1 store into another. I want to do this using EventFlow.MongoDB so that it sets up the databases/collections/events schema the way its expecting. I know there is support to replay the events, but that plays back into the read models. I'm sure I can dig into the source and figure out a way to do this by implementing interfaces somewhere. Probably will load all events using the EventStore store and save them all with a custom `IEventPersistence` for MongoDB that will retain all of the metadata, IDs, timestamps, etc. Unfortunately this is "low priority" and so I can only work on it when I find free time here and there 🙃

**Comment by warren-pieterse** [comment]
FYI I intend on migrating EventStore to v1 early in the new year. I've made
a draft PR that hasn't had any movement on as I've been busy with the
domain model and only using an in-memory event store for unit tests. Will
pick it up after the holidays...

**Comment by rasmus** [comment]
Does this change anything? Yet another NuGet rename? 😅 
https://www.kurrent.io/blog/kurrent-re-brand-faq

![image](https://github.com/user-attachments/assets/99501a6e-d855-453e-bf31-247c1b324018)

**Comment by warren-pieterse** [comment]
Oh wow, yes it does! I'll need to rebrand to align, and will reach out to
Rasmus in the new year about options. I'm not sure that supporting .net
standard is an option, I'd like to only implement a dot net v8+ version;
maybe create a completely new package aligned with the rebanded name that
reflects this is the best approach?

**Reply by rasmus** [reply]
Yes, `EventFlow.Kurrent` seems like a good idea. Ideally it should only be the namespace and name of the EventStore NuGet package that changes, not too much of the functionality within. Thus it should be possible to have most ready for when the rebrand is done.

**Comment by rasmus** [comment]
They are moving fast

![image](https://github.com/user-attachments/assets/b87357d4-a750-4ef1-9edf-2d18ae716239)
https://www.kurrent.io/blog/24.10-ga
